File: src/main/python/consumer/consumer.py
from swagger_client import XfxApi
from swagger_client.rest import ApiException

# Instantiating the XfxApi class
xfx = XfxApi()

while True:
    print("1. Browse remote shared folder\n2. Rename remote shared file or sub-folder \n3. Download remote shared file\n4. Upload local file to remote shared folder\n5. Delete shared remote file\n6. Exit")
    choice = int(input("Please enter your choice: "))
    if choice == 1:
        try:
            path = input(
                "Please enter the path of the root folder to browse all the files including sub-folders (SharedFolder/) or any specific sub-folder (SharedFolder/...): ")
            response = xfx.browse_files(path)
            print(response)
        except ApiException as ex:
            print("Exception occured when calling .browse_files().")
            print(ex)

    elif choice == 2:
        try:
            path = input(
                "Please enter the path of the file or folder you want to rename (ShardFolder/...): ")
            name = input(
                "Please enter the new name of the file or folder (name.txt for example): ")
            response = xfx.rename_file(path, name)
            print(response)
        except ApiException as ex:
            print("Exception occured when calling .rename_file().")
            print(ex)

    elif choice == 3:
        try:
            path = input(
                "Please enter the path of the file you would like to download (SharedFolder/...): ")
            response = xfx.download_file(path)
            # to get only the name of the file from the path to save it locally
            name = path.split("/")[-1]
            with open("ConsumerFolder/" + name, "wb") as file_out:
                file_out.write(eval(response))
                file_out.close()

            print("File saved successfully under: ConsumerFolder/" + name)
        except ApiException as ex:
            print("Exception occured when calling .download_file().")
            print(ex)

    elif choice == 4:
        try:
            target_path = input(
                "Please enter the path where you would like to upload to (SharedFolder/...): ")
            name = input(
                "Please enter the path of the local file you would like to upload (ConsumerFolder/...): ")
            response = xfx.upload_file(target_path, file=name)
            print(response)
        except ApiException as ex:
            print("Exception occured when calling .upload_file().")
            print(ex)
    elif choice == 5:
        try:
            path = input(
                "Please enter the path of the file you would like to delete (SharedFolder/...): ")
            response = xfx.delete_file(path)
            print(response)
        except ApiException as ex:
            print("Exception occured when calling .delete_file().")
            print(ex)

    elif choice == 6:
        print("Exiting... Thank you!")
        break

File: src/main/js/consumer.js
import OpenApiDefinition from './open_api_definition/src/index.js';
import * as fs from 'fs';
import readline from 'readline';


const xfxInstancePromise = Promise.resolve(new OpenApiDefinition.XfxApi());

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
});

const question = (prompt) => {
    return new Promise((resolve) => {
        rl.question(prompt, (answer) => {
            resolve(answer);
        });
    });
};

const browseFiles = async () => {
    return question(
        'Please enter the path of the root folder to browse all the files including sub-folders (SharedFolder/) or any specific sub-folder (SharedFolder/...): '
    )
        .then((path) => xfxInstancePromise.then((xfxInstance) => xfxInstance.browseFiles(path)))
        .then((response) => {
            console.log(`.browseFiles() was called successfully. The list of files: ${response}`);
            main();
        })
        .catch((ex) => {
            console.error('Exception occurred when calling .browseFiles(). ', ex);
            main();
        })
        .finally(() => {
            rl.close();
        });
};

const renameFile = async () => {
    return question('Please enter the path of the file to rename (SharedFolder/...): ')
        .then((path) => question('Please enter the new name of the file: ').then((newName) => [path, newName]))
        .then(([path, newName]) => xfxInstancePromise.then((xfxInstance) => xfxInstance.renameFile(path, newName)))
        .then((response) => {
            console.log(`.renameFile() was called successfully. The response: ${response}`);
            main();
        })
        .catch((ex) => {
            console.error('Exception occurred when calling .renameFile(). ', ex);
            main();
        })
};

const downloadFile = async (path) => {

    xfxInstancePromise.then((xfxInstance) => xfxInstance.downloadFile(path))
        .then((response) => {
            const fileName = path.split('/').pop();
            return new Promise((resolve, reject) => {
                fs.writeFile("JSConsumerFolder/" + fileName, response, (err) => {
                    if (err) {
                        reject(err);
                    } else {
                        resolve(`File saved successfully under: JSConsumerFolder/${fileName}`);
                    }
                });
            });
        })
        .then((message) => {
            console.log(message);
            main();
        })
        .catch((error) => {
            console.error(`Error downloading or saving file: ${error.message}`);
            main();
        })
};

const uploadFile = async (targetPath, localPath) => {

    xfxInstancePromise.then((xfxInstance) => {
        const file = fs.createReadStream(localPath);
        return new Promise((resolve, reject) => {

            xfxInstance.uploadFile(targetPath, { file: file })
                .then((response) => {
                    resolve(response);
                })
                .catch((ex) => {
                    reject(ex);
                });
        });
    })
        .then((response) => {
            console.log(
                `.uploadFile() was called successfully. The response: ${response}`
            );
            main();
        })
        .catch((ex) => {
            console.error("Exception occurred when calling .uploadFile(). ", ex);
            main();
        });
};

const deleteFile = async (path) => {
    xfxInstancePromise.then((xfxInstance) => xfxInstance.deleteFile(path))
        .then((response) => {
            console.log(`.deleteFile() was called successfully. The response: ${response}`);
            main();
        })
        .catch((ex) => {
            console.error('Exception occurred when calling .deleteFile(). ', ex);
            main();
        })
};


function main() {
    console.log(
        '1. Browse remote shared folder\n2. Rename remote shared file or sub-folder \n3. Download remote shared file\n4. Upload local file to remote shared folder\n5. Delete shared remote file\n6. Exit'
    );

    question('Please enter your choice: ').then((choice) => {
        const parsedChoice = parseInt(choice);

        if (parsedChoice === 1) {
            browseFiles();
        } else if (parsedChoice === 2) {
            renameFile();
        } else if (parsedChoice === 3) {
            question('Please enter the path of the file to download (SharedFolder/...): ')
                .then((path) => {
                    downloadFile(path);
                })
                .catch((error) => {
                    console.error(`Error processing user input: ${error.message}`);
                });
        } else if (parsedChoice === 4) {
            question('Please enter the path where you would like to upload to (SharedFolder/...): ')
                .then((targetPath) => {
                    question('Please enter the path of the file to upload (JSConsumerFolder/...): ')
                        .then((localPath) => {
                            uploadFile(targetPath, localPath);
                        })
                }).catch((error) => { console.error(`Error processing user input: ${error.message}`); });

        } else if (parsedChoice === 5) {
            question('Please enter the path of the file to delete (SharedFolder/...): ')
                .then((path) => {
                    deleteFile(path);
                })
                .catch((error) => {
                    console.error(`Error processing user input: ${error.message}`);
                });
        } else if (parsedChoice === 6) {
            console.log('Exiting... Thank you!');
            rl.close();
        } else {
            console.log('Invalid choice. Please try again.');
            main();
        }
    });
};

main();




